dlpy/attribute_utils.py

```diff
@@ -0,0 +1,763 @@
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+#
+
+''' Extended attribute table functions for the DLPy package '''
+
+import os
+import platform
+import numpy as np
+import pandas as pd
+import swat as sw
+import string
+import warnings
+import struct    
+
+def create_extended_attributes(conn, model_name, layers, data_spec, label_file_name=None):
+
+    '''
+    Create/override extended model attributes for given model
+
+    Update the extended model attributes given data spec(s).
+    The data spec(s) must define all relevant dictionary elements for
+    class :class:`DataSpec` or the resulting attribute table will
+    be inaccurate.
+
+    Parameters
+    ----------
+    conn : CAS
+        The CAS connection object
+    model_name : string
+        Specifies the name of the deep learning model
+    layers : list of strings
+        Specifies the names of all the layers in the deep learning
+        model
+    data_spec: list of :class:`DataSpec`
+        data specification for input and output layer(s)
+    label_file_name: string, optional
+        Fully qualified path to CSV file containing user-defined 
+        classification labels.  If not specified, numeric labels
+        are used.
+        
+    '''
+
+    # read user-supplied labels
+    if (label_file_name is None):
+        labels = None
+    else:
+        all_label_info = pd.read_csv(label_file_name, skipinitialspace=True, index_col=False)
+        labels = all_label_info['label'].values.tolist()
+    
+    # ensure table action loaded
+    rt = conn.queryactionset('table', _messagelevel = 'error')
+    if not rt:
+        conn.loadactionset('table', _messagelevel = 'error')
+        
+    # parse data spec(s) and create data spec attributes
+    ds_info = create_dataspec_attributes(conn, model_name, layers, data_spec)
+    
+    # update variable list attributes
+    create_varlist_attributes(conn, model_name, layers, ds_info)
+
+    # update variable information attributes
+    create_varinfo_attributes(conn, model_name, layers, ds_info, labels)
+
+    # update input parameter attributes
+    create_inputparm_attributes(conn, model_name, layers, ds_info)
+
+def create_dataspec_attributes(conn, model_name, layers, data_spec):
+
+    '''
+    Create/override extended model attributes for data spec(s)
+
+    Update the extended model attributes for the data spec(s).
+    The data spec must define all relevant dictionary elements for
+    class :class:`DataSpec` or the resulting attribute table will
+    be inaccurate.
+
+    Parameters
+    ----------
+    conn : CAS
```

Comment:
If that is true, then the headers for many functions will need to be changed throughout DLPy.  I just copied the header from a function in utils.py since I was using the conn object in exactly the same way.  I just did a quick check to verify that this object is used similarly in other files and has exactly the same description throughout.